File: apps/client/lib/features/auth/utils/auth_error.dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import '../../../core/l10n/l10n_ext.dart';

/// Parses a DioException from auth-service and returns a localized error string.
///
/// Handles the response shapes of the auth-error-codes contract:
///   1. Business body: `{ "code": "ACCOUNT_LOCKED", "params": { "minutes": 5 } }`
///      (NestJS replies with the thrown object verbatim — no `message` wrapper)
///   2. Validation body: `{ "message": ["VAL_EMAIL_INVALID", "VAL_PASSWORD_TOO_SHORT"] }`
///   3. Nested/legacy: `{ "message": { "code": ..., "params": ... } }` or a plain string
String authErrorToString(BuildContext context, DioException e) {
  final data = e.response?.data;
  if (data is Map) {
    // Business errors: `{ code, params }` at the top level.
    final topCode = data['code'];
    if (topCode is String && topCode.isNotEmpty) {
      final params = (data['params'] as Map?)?.cast<String, dynamic>();
      return _codeToString(context, topCode, params);
    }
    final msg = data['message'];
    if (msg is Map) {
      final code = msg['code'] as String?;
      final params = (msg['params'] as Map?)?.cast<String, dynamic>();
      if (code != null) return _codeToString(context, code, params);
    }
    if (msg is List && msg.isNotEmpty) {
      // Return the first validation error code
      return _codeToString(context, msg.first.toString(), null);
    }
    if (msg is String && msg.isNotEmpty) return msg; // legacy plain-string fallback
  }
  return context.l10n.errActionFailed;
}

String _codeToString(
  BuildContext context,
  String code,
  Map<String, dynamic>? params,
) {
  final l10n = context.l10n;
  switch (code) {
    // ── Success codes (shown as info/toast) ─────────────────────────────────
    case 'LOGIN_SUCCESS':
      return l10n.authMsgLoginSuccess;
    case 'LOGOUT_SUCCESS':
      return l10n.authMsgLogoutSuccess;
    case 'OTP_SENT':
      return l10n.authMsgOtpSent;
    case 'OTP_VALID':
      return l10n.authMsgOtpValid;
    case 'OTP_RESENT':
      return l10n.authMsgOtpResent;
    case 'PASSWORD_UPDATED':
      return l10n.authMsgPasswordUpdated;
    case 'REGISTER_SUCCESS':
      return l10n.authMsgRegisterSuccess;
    case 'ACCOUNT_UNVERIFIED_OTP_SENT':
      return l10n.authMsgAccountUnverifiedOtpSent;

    // ── OTP errors ───────────────────────────────────────────────────────────
    case 'OTP_INVALID':
      return l10n.authErrOtpInvalid;
    case 'OTP_EXPIRED':
      return l10n.authErrOtpExpired;
    case 'OTP_ATTEMPTS_EXCEEDED':
      return l10n.authErrOtpAttemptsExceeded;
    case 'OTP_WRONG_WITH_REMAINING':
      final remaining = _intParam(params, 'remaining');
      return l10n.authErrOtpWrongWithRemaining(remaining);
    case 'OTP_RESEND_COOLDOWN':
      final ttl = _intParam(params, 'ttl');
      return l10n.authErrOtpResendCooldown(ttl);

    // ── Email / validation errors ────────────────────────────────────────────
    case 'EMAIL_DOMAIN_INVALID':
      return l10n.authErrEmailDomainInvalid;
    case 'EMAIL_NOT_FOUND':
      return l10n.authErrEmailNotFound;
    case 'EMAIL_IN_USE':
      return l10n.authErrEmailInUse;
    case 'VAL_EMAIL_INVALID':
      return l10n.authErrValEmailInvalid;
    case 'VAL_EMAIL_REQUIRED':
      return l10n.authErrValEmailRequired;
    case 'VAL_DISPLAYNAME_REQUIRED':
      return l10n.authErrValDisplaynameRequired;
    case 'VAL_DISPLAYNAME_TOO_SHORT':
      return l10n.authErrValDisplaynameTooShort;
    case 'VAL_PASSWORD_TOO_SHORT':
      return l10n.authErrValPasswordTooShort;

    // ── Account / login errors ───────────────────────────────────────────────
    case 'ACCOUNT_LOCKED':
      final minutes = _intParam(params, 'minutes');
      return l10n.authErrAccountLocked(minutes);
    case 'LOGIN_FAILED_WITH_REMAINING':
      final remaining = _intParam(params, 'remaining');
      return l10n.authErrLoginFailedWithRemaining(remaining);
    case 'LOGIN_FAILED_LOCKED':
      final minutes = _intParam(params, 'minutes');
      return l10n.authErrLoginFailedLocked(minutes);

    // ── Token / session errors ───────────────────────────────────────────────
    case 'TOKEN_INVALID':
      return l10n.authErrTokenInvalid;
    case 'SESSION_NOT_FOUND':
      return l10n.authErrSessionNotFound;
    case 'SESSION_INVALID':
      return l10n.authErrSessionInvalid;
    case 'SESSION_REVOKED':
      return l10n.authErrSessionRevoked;
    case 'REFRESH_TOKEN_REUSE':
      return l10n.authErrRefreshTokenReuse;
    case 'REFRESH_TOKEN_INVALID':
      return l10n.authErrRefreshTokenInvalid;
    case 'REFRESH_TOKEN_ROTATED':
      return l10n.authErrRefreshTokenRotated;
    case 'TOKEN_SESSION_MISMATCH':
      return l10n.authErrTokenSessionMismatch;

    // ── Social / misc errors ─────────────────────────────────────────────────
    case 'SOCIAL_EMAIL_UNAVAILABLE':
      return l10n.authErrSocialEmailUnavailable;
    case 'LOGIN_CODE_INVALID':
      return l10n.authErrLoginCodeInvalid;
    case 'USER_NOT_FOUND':
      return l10n.authErrUserNotFound;

    default:
      return l10n.errActionFailed;
  }
}

/// Safely extracts an int param — returns 0 if missing or wrong type.
int _intParam(Map<String, dynamic>? params, String key) {
  if (params == null) return 0;
  final v = params[key];
  if (v is int) return v;
  if (v is num) return v.toInt();
  return 0;
}

File: apps/web/lib/auth/auth-error.ts
import { AxiosError } from 'axios'

interface AuthErrorBody {
  code: string
  params?: Record<string, string | number>
}

/** Extract { code, params } from an AxiosError thrown by auth-service. */
export function parseAuthError(err: unknown): { code: string; params?: Record<string, string | number> } {
  const e = err as AxiosError<{ code?: string; params?: Record<string, string | number>; message?: unknown }>
  const data = e?.response?.data
  // Business errors: NestJS replies with the thrown object verbatim, so the
  // body is `{ code, params }` at the top level (no `message` wrapper).
  if (data && typeof data.code === 'string' && data.code) {
    return { code: data.code, params: data.params }
  }
  const msg = data?.message
  // Defensive: also accept a nested `{ message: { code, params } }` shape.
  if (typeof msg === 'object' && msg !== null && !Array.isArray(msg)) {
    const body = msg as AuthErrorBody
    if (body.code) return { code: body.code, params: body.params }
  }
  // Validation errors (class-validator) arrive as `{ message: ["VAL_..."] }`.
  if (Array.isArray(msg) && msg.length > 0) {
    return { code: String(msg[0]) }
  }
  return { code: 'GENERIC_ERROR' }
}

/** Map an auth-service error code to a next-intl translation key path. */
export function authCodeToI18nKey(code: string): string {
  const map: Record<string, string> = {
    // Success codes
    LOGIN_SUCCESS: 'msgLoginSuccess',
    LOGOUT_SUCCESS: 'msgLogoutSuccess',
    OTP_SENT: 'msgOtpSent',
    OTP_VALID: 'msgOtpValid',
    OTP_RESENT: 'msgOtpResent',
    PASSWORD_UPDATED: 'msgPasswordUpdated',
    REGISTER_SUCCESS: 'msgRegisterSuccess',
    ACCOUNT_UNVERIFIED_OTP_SENT: 'msgAccountUnverifiedOtpSent',
    // Error codes
    OTP_INVALID: 'errOtpInvalid',
    OTP_EXPIRED: 'errOtpExpired',
    OTP_ATTEMPTS_EXCEEDED: 'errOtpAttemptsExceeded',
    OTP_WRONG_WITH_REMAINING: 'errOtpWrongWithRemaining',
    OTP_RESEND_COOLDOWN: 'errOtpResendCooldown',
    EMAIL_DOMAIN_INVALID: 'errEmailDomainInvalid',
    ACCOUNT_LOCKED: 'errAccountLocked',
    LOGIN_FAILED_WITH_REMAINING: 'errLoginFailedWithRemaining',
    LOGIN_FAILED_LOCKED: 'errLoginFailedLocked',
    TOKEN_INVALID: 'errTokenInvalid',
    SESSION_NOT_FOUND: 'errSessionNotFound',
    SESSION_INVALID: 'errSessionInvalid',
    SESSION_REVOKED: 'errSessionRevoked',
    REFRESH_TOKEN_REUSE: 'errRefreshTokenReuse',
    REFRESH_TOKEN_INVALID: 'errRefreshTokenInvalid',
    REFRESH_TOKEN_ROTATED: 'errRefreshTokenRotated',
    TOKEN_SESSION_MISMATCH: 'errTokenSessionMismatch',
    SOCIAL_EMAIL_UNAVAILABLE: 'errSocialEmailUnavailable',
    LOGIN_CODE_INVALID: 'errLoginCodeInvalid',
    EMAIL_NOT_FOUND: 'errEmailNotFound',
    USER_NOT_FOUND: 'errUserNotFound',
    EMAIL_IN_USE: 'errEmailInUse',
    // Validation codes
    VAL_EMAIL_INVALID: 'errValEmailInvalid',
    VAL_EMAIL_REQUIRED: 'errValEmailRequired',
    VAL_DISPLAYNAME_REQUIRED: 'errValDisplayNameRequired',
    VAL_DISPLAYNAME_TOO_SHORT: 'errValDisplayNameTooShort',
    VAL_PASSWORD_TOO_SHORT: 'errValPasswordTooShort',
    // Fallback
    GENERIC_ERROR: 'errGeneric',
  }
  return map[code] ?? 'errGeneric'
}
